doc/fwutil/fwutil.md

```diff
@@ -0,0 +1,441 @@
+|:-------|:----------------------------------------------------|
+| FW     | Firmware                                            |
+| SONiC  | Software for Open Networking in the Cloud           |
+| PSU    | Power Supply Unit                                   |
+| QSFP   | Quad Small Form-factor Pluggable                    |
+| EEPROM | Electrically Erasable Programmable Read-Only Memory |
+| I2C    | Inter-Integrated Circuit                            |
+| SPI    | Serial Peripheral Interface                         |
+| JTAG   | Joint Test Action Group                             |
+| BIOS   | Basic Input/Output System                           |
+| CPLD   | Complex Programmable Logic Device                   |
+| FPGA   | Field-Programmable Gate Array                       |
+| URL    | Uniform Resource Locator                            |
+| API    | Application Programming Interface                   |
+| N/A    | Not Applicable/Not Available                        |
+
+## List of figures
+
+[Figure 1: FW utility High Level Design](#figure-1-fw-utility-high-level-design)  
+[Figure 2: Show components status flow](#figure-2-show-components-status-flow)  
+[Figure 3: FW install (non modular) flow](#figure-3-fw-install-non-modular-flow)  
+[Figure 4: FW install (modular) flow](#figure-4-fw-install-modular-flow)  
+
+## List of tables
+
+[Table 1: Event logging](#table-1-event-logging)
+
+# 1 Introduction
+
+## 1.1 Feature overview
+
+A modern network switch is a sophisticated equipment which consists of many auxiliary components  
+which are responsible for managing different subsystems (e.g., PSU/FAN/QSFP/EEPROM/THERMAL)  
+and providing necessary interfaces (e.g., I2C/SPI/JTAG).
+
+Basically these components are complex programmable logic devices with it's own HW architecture  
+and software. The most important are BIOS/CPLD/FPGA etc.
+
+It is very important to always have the latest recommended software version to improve device stability,  
+security and performance. Also, software updates can add new features and remove outdated ones.
+
+In order to make software update as simple as possible and to provide a nice user frindly  
+interface for various maintenance operations (e.g., install a new FW or query current version)  
+we might need a dedicated FW utility.
+
+## 1.2 Requirements
+
+### 1.2.1 Functionality
+
+**This feature will support the following functionality:**
+1. Manual FW installation for particular platform component
+2. Automatic FW installation for all available platform components
+3. Querying platform components and FW versions
+
+### 1.2.2 Command interface
+
+**This feature will support the following commands:**
+1. show: display FW versions
+2. install: manual FW installation
+3. update: automatic FW installation
+
+### 1.2.3 Error handling
+
+**This feature will provide error handling for the next situations:**
+1. Invalid input
+2. Incompatible options/parameters
+3. Invalid/nonexistent FW URL/path
```

Comment:
Can design supports images pull from outside the device? (remote URLs over HTTP/HTTPS)?

Reply:
Yes. The example is local path but you can use URL. This applies only to the manual mode and not the automatic which was also discussed but not yet finalized.